File: StriverA2Z/Java/LinkedLists/Questions/SinglyLinkedList/RemoveNElementFromEnd.java
public class RemoveNElementFromEnd {
    public static void main(String[] args) {
        Node head = LinkedListOperations.arrayToLinkedList(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9});
        head = removeElementFromEnd(head, 1);
        LinkedListOperations.traverseLinkedList(head);
    }

    static Node removeElementFromEnd(Node head, int N) {
        Node fast = head;
        for (int i = 0; i < N; i++) fast = fast.next;
        if (fast == null) return head.next;
        Node slow = head;

        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        slow.next = slow.next.next;
        return head;
    }
}

File: StriverA2Z/C++/6_LinkedList/Questions/SinglyLinkedList/remove_n_element_from_end.cpp
#include<bits/stdc++.h>
#include "Node.h";

using namespace std;

Node *removeElementFromEnd(Node *head, int N) {
    Node *fast = head;
    for (int i = 0; i < N; i++) fast = fast->next;
    if (fast == nullptr) {
        Node *newHead = head->next;
        delete head;
        return newHead;
    }
    Node *slow = head;

    while (fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next;
    }
    Node *delNode = slow->next;
    slow->next = slow->next->next;
    delete delNode;
    return head;
}

int main() {
    vector<int> vec = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    Node *head = array_to_linkedlist(vec);
    head = removeElementFromEnd(head, 10);
    traverseLinkedList(head);
    return 0;
}